packages/firestore/ios/Plugin/FirebaseFirestore.swift

```diff
@@ -278,26 +302,19 @@ import FirebaseFirestore
                         completion(result, nil)
                     }
                 }
-                self.listenerRegistrationMap[callbackId] = listenerRegistration
+                await listenerRegistrationMap.addRegistration(listenerRegistration, listenerId: callbackId)
             } catch {
                 completion(nil, error)
             }
         }
     }
 
-    @objc public func removeSnapshotListener(_ options: RemoveSnapshotListenerOptions) {
+    @objc public func removeSnapshotListener(_ options: RemoveSnapshotListenerOptions) async {
         let callbackId = options.getCallbackId()
-
-        if let listenerRegistration = self.listenerRegistrationMap[callbackId] {
-            listenerRegistration.remove()
-        }
-        self.listenerRegistrationMap.removeValue(forKey: callbackId)
+        await listenerRegistrationMap.removeRegistration(listenerId: callbackId)
     }
 
-    @objc public func removeAllListeners() {
-        for listenerRegistration in self.listenerRegistrationMap.values {
-            listenerRegistration.remove()
-        }
-        self.listenerRegistrationMap.removeAll()
+    @objc public func removeAllListeners() async {
```

Comment:
```suggestion
    @objc public func removeAllListeners(completion: @escaping () -> Void) async {
```

Please use a completion handler instead.